src/engines/join-propagation.ts

```diff
@@ -0,0 +1,251 @@
+  if (distinctHeadVariables.some(x => x < 0 || x >= nodes.length)) throwErr('Some of the head variables do not exist in the network.')
+  if (distinctParentVariables.some(x => x < 0 || x >= nodes.length)) throwErr('Some of the parent variables do not exist in the network.')
+  const joinDomain = [...headVariables, ...parentVariables]
+
+  // Create a copy of the formulas that we can use for performing the traversal.   This collection of formulas
+  // may be updated during message passing with new formulas that were not part of the original
+  // junction tree message passing.
+  const amendedFormulas = [...formulas]
+
+  // Create a list of connected components that have some of the variables of interest.
+  // Each of these connected components will need to be traversed to compute the joint
+  // distribution.
+  const connectedComponentsThatHaveVariablesOfInterest = connectedComponents.filter(cc =>
+    cc.some(x => cliques[x].domain.some(i => joinDomain.includes(i))),
+  )
+
+  // Initialize two dictionaries that will be used by the message passing
+  // algorithm to find existing formulas and messages, and avoid duplication
+  // of effort.
+  const messages: Record<string, Formula[]> = {}
+  const fmap: Record<string, number> = {}
+  formulas.forEach((f, i) => {
+    fmap[f.name] = i
+  })
+  // This helper function will be used during message passing to add new formulas
+  // to the dictionary, but only if it doesn't already exist.
+  const upsert = upsertFormula(amendedFormulas, fmap)
+
+  // traverse each connected component, collecting all the messages into
+  // a root clique.   This traversal will return the formula for the join of
+  // all the variables that occur in each clique.
+  const cliqueFormulas = connectedComponentsThatHaveVariablesOfInterest.map(cc => {
+    // get the cliques and variables that are in this connected component.
+    const theseCliques = cc.map(i => cliques[i])
+    const theseVariables = joinDomain.filter(i => theseCliques.some(c => c.domain.includes(i)))
+
+    // find a best root clique into which the messages will be collected
+    const rootClique = pickRootClique(theseCliques, theseVariables, amendedFormulas)
+
+    // recursively collect the evidence from the neihgbor cliques.   This message passing
+    // has been modified from the one used for making the clique graph consistent in that
+    // it does not marginalize out the variables in the join domain.  The clique evidence
+    // does not need to be distributed out to the other nodes because they are already
+    // consistent.
+    collectCliquesEvidence(cliques, separators, messages, upsert, amendedFormulas, theseVariables, rootClique.id)
+    const messagesReceived: Formula[][] = rootClique.neighbors.map(x => messages[messageName(x, rootClique.id)] || [])
+    // after receiving the messages, we need to multiply them together with the root clique's
```

Comment:
nit: Sentence needs to start with a capital.

Reply:
fixed